src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToComponentCodeActionProvider.cs

```diff
@@ -0,0 +1,214 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using ICSharpCode.Decompiler.CSharp.Syntax;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+using Microsoft.AspNetCore.Razor.LanguageServer.CodeActions.Models;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.AspNetCore.Razor.Threading;
+using Microsoft.CodeAnalysis.Razor.Logging;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.CodeActions.Razor;
+
+internal sealed class ExtractToComponentCodeActionProvider() : IRazorCodeActionProvider
+{
+    public Task<ImmutableArray<RazorVSInternalCodeAction>> ProvideAsync(RazorCodeActionContext context, CancellationToken cancellationToken)
+    {
+        if (context is null)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        if (!context.SupportsFileCreation)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        if (!FileKinds.IsComponent(context.CodeDocument.GetFileKind()))
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        var syntaxTree = context.CodeDocument.GetSyntaxTree();
+        if (syntaxTree?.Root is null)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
+        // Make sure the selection starts on an element tag
+        var (startNode, endNode) = GetStartAndEndElements(context, syntaxTree);
+        if (startNode is null || endNode is null)
+        {
+            return SpecializedTasks.EmptyImmutableArray<RazorVSInternalCodeAction>();
+        }
+
```

Comment:
Shouldn't there be code here that doesn't offer the action if there is C# code in the selection? Or are we just going with the approach of offering to do _something_ even if its not great?

Reply:
I believe the latter, but @leslierichardson95 what do you think?

Reply:
It's fine if that is the determination. Though would love to see tests that have it. In particular I'm curious about things like:

```
<div>blah</div>

[|Hello

@code {
  |]
}
```

or:
```
<div>blah</div>

[|Hello

@if (true) {
	<div>  |] </div>
}
```

or the very exciting

```
<div>blah</div>

[|Hello

 @{
     RenderFragment fragment =
     @<Component1 Id="Comp1"
                  Caption="Title">|]
     </Component1>;
 }
```

Reply:
yea, I left your comment open on tests. Adding more today for sure

Reply:
> I believe the latter, but @leslierichardson95 what do you think?

Last time we chatted, I think we decided on the latter of showing the option which I'm still ok with as long as we still have telemetry attached that lets us generally know what the user was trying to accomplish.